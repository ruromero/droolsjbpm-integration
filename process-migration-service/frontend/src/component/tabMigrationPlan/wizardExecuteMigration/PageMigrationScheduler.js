import React from "react";
import PropTypes from "prop-types";

import Datetime from "react-datetime";
import validator from "validator";
import moment from "moment";
import FormGroup from "patternfly-react/dist/js/components/Form/FormGroup";
import {
  ControlLabel,
  FormControl,
  HelpBlock
} from "patternfly-react/dist/js/components/Form";
import Radio from "patternfly-react/dist/js/components/Form/Radio";

export default class PageMigrationScheduler extends React.Component {
  constructor(props) {
    super(props);
    const isValidUrl = this.validateCallbackUrl(props.callbackUrl) === null;
    const isValidTime =
      this.validateScheduledStartTime(
        this.formatScheduledStartTime(props.scheduledStartTime)
      ) === null;
    this.state = {
      scheduleMigration: this.props.scheduledStartTime !== "",
      isValidUrl,
      isValidTime,
      showTimePicker: false
    };
    this.onValidationChange(isValidUrl, isValidTime);
  }

  onValidationChange = (isValidUrl, isValidTime) => {
    this.props.onIsValid(isValidUrl && isValidTime);
  };

  validateCallbackUrl = callbackUrl => {
    let validationMessageUrl = null;
    if (
      callbackUrl !== undefined &&
      callbackUrl !== "" &&
      !validator.isURL(callbackUrl, { require_tld: false })
    ) {
      validationMessageUrl = "error";
    }
    return validationMessageUrl;
  };

  validateScheduledStartTime = scheduledStartTime => {
    if (scheduledStartTime === null || moment().diff(scheduledStartTime) < 0) {
      return null;
    }
    return "error";
  };

  changeCallbackUrl = event => {
    const inputUrl = event.target.value;
    const isValidUrl = this.validateCallbackUrl(inputUrl) === null;
    this.props.onFieldChange("callbackUrl", inputUrl === "" ? null : inputUrl);
    this.setState({ isValidUrl });
    this.onValidationChange(isValidUrl, this.state.isValidTime);
  };

  handleDateTimeInput = inputMoment => {
    const validationMessageTime = this.validateScheduledStartTime(inputMoment);
    const isValidTime = validationMessageTime === null;
    if (isValidTime) {
      this.props.onFieldChange("scheduledStartTime", inputMoment.format());
    }
    this.setState({ isValidTime, validationMessageTime });
    this.onValidationChange(this.state.isValidUrl, isValidTime);
  };

  isDateSelectable = current =>
    current.isAfter(Datetime.moment().subtract(1, "day"));

  hideScheduleMigration = () => {
    this.setState({
      scheduleMigration: false
    });
    this.props.onFieldChange("scheduledStartTime", null);
  };

  formatScheduledStartTime = scheduledStartTime => {
    if (scheduledStartTime === "") {
      return null;
    }
    return moment(this.props.scheduledStartTime, "YYYY-MM-DDTHH:mm:ssZ", true);
  };

  render() {
    const scheduledStartTime = this.formatScheduledStartTime(
      this.props.scheduledStartTime
    );
    const callbackUrlValidation = this.validateCallbackUrl(
      this.props.callbackUrl
    );
    const scheduledStartTimeValidation = this.validateScheduledStartTime(
      scheduledStartTime
    );

    return (
      <React.Fragment>
        <FormGroup
          controlId="formMigrationScheduler_CallbackURL"
          validationState={callbackUrlValidation}
        >
          <ControlLabel>Callback URL</ControlLabel>
          <FormControl
            type="text"
            value={this.props.callbackUrl}
            onChange={this.changeCallbackUrl}
          />
          <FormControl.Feedback />
          {callbackUrlValidation && <HelpBlock>Enter a valid URL</HelpBlock>}
        </FormGroup>
        <FormGroup
          controlId="formMigrationScheduler_Schedule"
          validationState={scheduledStartTimeValidation}
        >
          <ControlLabel>Run migration</ControlLabel>
          <FormGroup controlId="formMigrationScheduler_ScheduleRadio">
            <Radio
              checked={!this.state.scheduleMigration}
              onChange={this.hideScheduleMigration}
            >
              Now
            </Radio>
            <Radio
              checked={this.state.scheduleMigration}
              onChange={() => this.setState({ scheduleMigration: true })}
            >
              Schedule
            </Radio>
          </FormGroup>
          {this.state.scheduleMigration && (
            <Datetime
              id="PageMigrationScheduler_scheduleTime"
              onChange={this.handleDateTimeInput}
              isValidDate={this.isDateSelectable}
              value={scheduledStartTime}
              input
              onFocus={() => this.setState({ showTimePicker: true })}
              onBlur={() => this.setState({ showTimePicker: false })}
              open={this.state.showTimePicker}
            />
          )}
          <FormControl.Feedback />
          {scheduledStartTimeValidation && (
            <HelpBlock>Select a valid date in the future</HelpBlock>
          )}
        </FormGroup>
      </React.Fragment>
    );
  }
}

PageMigrationScheduler.defaultProps = {
  callbackUrl: "",
  scheduledStartTime: ""
};

PageMigrationScheduler.propTypes = {
  callbackUrl: PropTypes.string,
  scheduledStartTime: PropTypes.string,
  onFieldChange: PropTypes.func.isRequired,
  onIsValid: PropTypes.func.isRequired
};
