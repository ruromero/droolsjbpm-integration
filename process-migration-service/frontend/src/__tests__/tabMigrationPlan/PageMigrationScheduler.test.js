import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";

import PageMigrationScheduler from "../../component/tabMigrationPlan/wizardExecuteMigration/PageMigrationScheduler.js";

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

test("PageMigrationScheduler renders expected components", () => {
  const callbackFn = jest.fn();
  const isValidFn = jest.fn();
  const { container, getByLabelText } = render(
    <PageMigrationScheduler onFieldChange={callbackFn} onIsValid={isValidFn} />
  );

  expect(container).toMatchSnapshot();
  const callbackInput = getByLabelText("Callback URL");
  expect(callbackInput).toHaveProperty("value", "");

  fireEvent.change(callbackInput, {
    target: { value: "some url" }
  });

  expect(callbackFn).toHaveBeenCalled();
  expect(callbackFn.mock.calls[0][1]).toBe("some url");
});

test("PageMigrationScheduler renders error for wrong url", () => {
  const callbackFn = jest.fn();
  const isValidFn = jest.fn();
  const { container, getByLabelText } = render(
    <PageMigrationScheduler
      callbackUrl="wrong url"
      onFieldChange={callbackFn}
      onIsValid={isValidFn}
    />
  );

  expect(container).toMatchSnapshot();
  const callbackInput = getByLabelText("Callback URL");
  expect(callbackInput).toHaveProperty("value", "wrong url");
});

test("PageMigrationScheduler renders callback url", () => {
  const callbackFn = jest.fn();
  const isValidFn = jest.fn();
  const callbackUrl = "https://example.com/callback";
  const { container, getByLabelText } = render(
    <PageMigrationScheduler
      callbackUrl={callbackUrl}
      onFieldChange={callbackFn}
      onIsValid={isValidFn}
    />
  );

  expect(container).toMatchSnapshot();
  const callbackInput = getByLabelText("Callback URL");
  expect(callbackInput).toHaveProperty("value", callbackUrl);
});

test("PageMigrationScheduler is scheduled", () => {
  const callbackFn = jest.fn();
  const isValidFn = jest.fn();
  const callbackUrl = "https://example.com/callback";
  const scheduledTime = "2019-7-17T20:59:22.128Z";
  const { getByLabelText } = render(
    <PageMigrationScheduler
      callbackUrl={callbackUrl}
      scheduledStartTime={scheduledTime}
      onFieldChange={callbackFn}
      onIsValid={isValidFn}
    />
  );

  expect(getByLabelText("Schedule")).toHaveProperty("checked", true);
});
