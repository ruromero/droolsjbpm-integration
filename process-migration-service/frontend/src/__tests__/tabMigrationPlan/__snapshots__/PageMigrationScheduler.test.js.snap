// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMigrationScheduler renders callback url 1`] = `
<div>
  <div
    class="form-group has-feedback"
  >
    <label
      class="control-label"
      for="formMigrationScheduler_CallbackURL"
    >
      Callback URL
    </label>
    <input
      class="form-control"
      id="formMigrationScheduler_CallbackURL"
      type="text"
      value="https://example.com/callback"
    />
  </div>
  <div
    class="form-group has-feedback"
  >
    <label
      class="control-label"
      for="formMigrationScheduler_Schedule"
    >
      Run migration
    </label>
    <div
      class="form-group"
    >
      <div
        class="radio"
      >
        <label
          title=""
        >
          <input
            checked=""
            type="radio"
          />
          Now
        </label>
      </div>
      <div
        class="radio"
      >
        <label
          title=""
        >
          <input
            type="radio"
          />
          Schedule
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageMigrationScheduler renders error for wrong url 1`] = `
<div>
  <div
    class="form-group has-feedback has-error"
  >
    <label
      class="control-label"
      for="formMigrationScheduler_CallbackURL"
    >
      Callback URL
    </label>
    <input
      class="form-control"
      id="formMigrationScheduler_CallbackURL"
      type="text"
      value="wrong url"
    />
    <span
      class="form-control-feedback glyphicon glyphicon-remove"
    />
    <span
      class="help-block"
    >
      Enter a valid URL
    </span>
  </div>
  <div
    class="form-group has-feedback"
  >
    <label
      class="control-label"
      for="formMigrationScheduler_Schedule"
    >
      Run migration
    </label>
    <div
      class="form-group"
    >
      <div
        class="radio"
      >
        <label
          title=""
        >
          <input
            checked=""
            type="radio"
          />
          Now
        </label>
      </div>
      <div
        class="radio"
      >
        <label
          title=""
        >
          <input
            type="radio"
          />
          Schedule
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`PageMigrationScheduler renders expected components 1`] = `
<div>
  <div
    class="form-group has-feedback"
  >
    <label
      class="control-label"
      for="formMigrationScheduler_CallbackURL"
    >
      Callback URL
    </label>
    <input
      class="form-control"
      id="formMigrationScheduler_CallbackURL"
      type="text"
      value=""
    />
  </div>
  <div
    class="form-group has-feedback"
  >
    <label
      class="control-label"
      for="formMigrationScheduler_Schedule"
    >
      Run migration
    </label>
    <div
      class="form-group"
    >
      <div
        class="radio"
      >
        <label
          title=""
        >
          <input
            checked=""
            type="radio"
          />
          Now
        </label>
      </div>
      <div
        class="radio"
      >
        <label
          title=""
        >
          <input
            type="radio"
          />
          Schedule
        </label>
      </div>
    </div>
  </div>
</div>
`;
